ADBDEV-4904-28: Add check that prefix is not NULL to the patternsel function

red1452 opened the pull request:
Add check that prefix is not NULL to the patternsel function

The prefix pointer can be NULL at the patternsel function in case of it was not
set at the pattern_fixed_prefix function.

Insist was added for checking this situation, since prefix can be NULL only if
pstatus is Pattern_Prefix_None.

red1452 committed:
Add check that prefix is not NULL to the patternsel function

The prefix pointer can be NULL at the patternsel function in case of it was not
set at the pattern_fixed_prefix function. Assertion was added for checking this
situation.

RekGRpth commented on src/backend/utils/adt/selfuncs.c:
```diff
@@ -1281,6 +1281,8 @@ patternsel(PG_FUNCTION_ARGS, Pattern_Type ptype, bool negate)
 			Selectivity heursel;
 			Selectivity prefixsel;
 
+			Assert(prefix);
```
I think that instead of asserting here, you should have removed the condition here https://github.com/arenadata/gpdb/blob/fd5939d4f491238bbd28be88b98949cd0e2d6fe0/src/backend/utils/adt/selfuncs.c#L1211 and there https://github.com/arenadata/gpdb/blob/fd5939d4f491238bbd28be88b98949cd0e2d6fe0/src/backend/utils/adt/selfuncs.c#L1342, and also added an assertion before the removed first condition.

RekGRpth committed:
Add check that prefix is not NULL to the patternsel function

The prefix pointer can be NULL at the patternsel function in case of it was not
set at the pattern_fixed_prefix function. Insist was added for checking this
situation.

RekGRpth committed:
additional fix

RekGRpth commented:
> > Insist was added for checking this situation.
> 
> Can you change to:
> 
> > Insist was added for checking this situation, since prefix can be NULL only if
> > pstatus is Pattern_Prefix_None.

updated

bandetto commented on src/backend/utils/adt/selfuncs.c:
```diff
@@ -1241,6 +1241,8 @@ patternsel(PG_FUNCTION_ARGS, Pattern_Type ptype, bool negate)
 		if (eqopr == InvalidOid)
 			elog(ERROR, "no = operator for opfamily %u", opfamily);
+
+		Insist(prefix != NULL);
```
https://github.com/arenadata/gpdb/blob/cbef1396b38e20a82ba358b157445e4d673fc7ff/src/backend/utils/adt/selfuncs.c#L1205-L1206

Maybe we should just do this right under the `pattern_fixed_prefix()` call:
```c
		AssertImply(pstatus != Pattern_Prefix_None, prefix != NULL);
```
?

RekGRpth replied:
Yes, but this will not save us from the static analyzer.

bandetto replied:
Can we get rid of the warning with this instead of each `Insist()`?
```c
if (pstatus != Pattern_Prefix_None)
	Insist(prefix != NULL);
```

RekGRpth replied:
I think, no, because [condition](https://github.com/arenadata/gpdb/pull/770/files#diff-60407b5c46aaf9ea36ed1002b1bfaf5e818298c30d3d680c4e9c91f814befd46R1211) was left
